File: matlab/+mestra/Callable.m
classdef (Abstract) Callable
%mestra.Callable  The four-method callable protocol.
%
%   A callable maps keys in to predictions out.  Specification section
%   10 fixes it at four things and nothing more:
%
%       call       a keys table in, one prediction per output out
%       toDict     a nested dictionary that fully represents it
%       fromDict   the inverse, a static method dispatched on the type
%                  string through mestra.Registry
%       disp       optional, one line for printing
%
%   Everything else a callable knows is inside its dictionary and is
%   its own business.  The format does not constrain it.
%
%   The keys table is a MATLAB table whose variable names are the key
%   names (section 26).  call returns a containers.Map from the name
%   of an output to a prediction: a struct with the fields
%
%       mean         a mestra.Array shaped as the slot is stored, with
%                    the file's own axis order: (row, node | cell,
%                    component) for an array slot and (row) for a
%                    scalar slot
%       uncertainty  a band (section 9), the half-width of the
%                    interval around the mean, of the same shape, or
%                    [] when the model has none
%       level        the coverage the band claims, in (0, 1), or []
%       method       how the band was made, one sentence, or ''
%
%   Build one with mestra.Callable.prediction, which checks that the
%   three band fields come together and that the band is never
%   negative.  How the band was computed is the model's business; what
%   it claims is on the record.
%
%   To add a type, subclass this, implement the three methods, and
%   register it:
%
%       mestra.Registry.register('my_type', @MyType.fromDict);
%
%   See also mestra.Affine, mestra.Registry, mestra.evaluate,
%   mestra.prediction.

    methods (Abstract)
        % out = call(obj, keysTable)
        out = call(obj, keysTable)

        % d = toDict(obj)  -- a containers.Map, per the codec
        d = toDict(obj)
    end

    methods (Abstract, Static)
        % obj = fromDict(d)
        obj = fromDict(d)
    end

    methods
        function s = repr(obj) %#ok<MANU>
        %repr  One line describing the callable.  Optional; override it.
            s = '';
        end

        function t = type(obj)
        %type  The public type string, taken from the registry.
            t = mestra.Registry.typeOf(class(obj));
        end

        function disp(obj)
        %disp  One line naming the type and the description.
            r = obj.repr();
            if isempty(r)
                fprintf('  %s callable\n', obj.type());
            else
                fprintf('  %s\n', r);
            end
        end
    end

    methods (Static)
        function p = prediction(meanValue, uncertainty, level, method)
        %prediction  The record a callable returns for one output.
        %   P = mestra.Callable.prediction(MEAN) is a prediction with
        %   no band.  P = mestra.Callable.prediction(MEAN, UNCERTAINTY,
        %   LEVEL, METHOD) carries one: UNCERTAINTY is a mestra.Array
        %   of the mean's shape and never negative, LEVEL the coverage
        %   it claims in (0, 1), and METHOD one sentence on how it was
        %   made.  A band missing any of the three is refused here
        %   rather than in a file (section 10).
            if nargin < 2, uncertainty = []; end
            if nargin < 3, level = []; end
            if nargin < 4, method = ''; end
            if ~isa(meanValue, 'mestra.Array')
                error('mestra:prediction', ...
                      ['the mean of a prediction is a mestra.Array ' ...
                       'shaped as the slot is stored (section 10)']);
            end
            if isstring(method), method = char(method); end
            if isempty(uncertainty)
                if ~isempty(level) || ~isempty(method)
                    error('mestra:prediction', ...
                          ['level and method go with an uncertainty; a ' ...
                           'prediction without one has neither']);
                end
                p = struct('mean', meanValue, 'uncertainty', [], ...
                           'level', [], 'method', '');
                return
            end
            if ~isa(uncertainty, 'mestra.Array')
                error('mestra:prediction', ...
                      'the uncertainty of a prediction is a mestra.Array');
            end
            if ~isequal(uncertainty.shape, meanValue.shape)
                error('mestra:prediction', ...
                      ['a band has the shape of its mean; the mean is ' ...
                       '%s and the band %s'], mat2str(meanValue.shape), ...
                      mat2str(uncertainty.shape));
            end
            if isempty(level) || ~isnumeric(level) || ~isscalar(level) ...
                    || ~(level > 0 && level < 1)
                error('mestra:prediction', ...
                      ['a band states the coverage it claims as level in ' ...
                       '(0, 1); a 1.96-sigma Gaussian band is 0.95']);
            end
            if ~ischar(method) || isempty(method)
                error('mestra:prediction', ...
                      'a band says how it was made; give method one sentence');
            end
            if any(double(uncertainty.data(:)) < 0)
                error('mestra:prediction', ...
                      'a band is a half-width and is never negative');
            end
            p = struct('mean', meanValue, 'uncertainty', uncertainty, ...
                       'level', double(level), 'method', method);
        end
    end
end

File: cpp/src/callable.cpp
#include "mestra/callable.hpp"

#include <map>

#include "mestra/io.hpp"

namespace mestra {

std::size_t KeysTable::rows() const {
  for (const std::vector<double>& c : numeric) {
    if (!c.empty()) return c.size();
  }
  for (const std::vector<std::string>& c : text) {
    if (!c.empty()) return c.size();
  }
  // Every column is empty: the table has the length the columns agree
  // on, which is zero.
  return 0;
}

bool KeysTable::has(const std::string& name) const {
  for (const std::string& n : names) {
    if (n == name) return true;
  }
  return false;
}

const std::vector<double>& KeysTable::column(const std::string& name) const {
  for (std::size_t i = 0; i < names.size(); ++i) {
    if (names[i] == name && i < numeric.size()) return numeric[i];
  }
  // Section 26: a missing column that the callable declares is an
  // error at call time.
  throw Error("", "the keys table has no column \"" + name + "\"");
}

const std::vector<std::string>& KeysTable::text_column(
    const std::string& name) const {
  for (std::size_t i = 0; i < names.size(); ++i) {
    if (names[i] == name && i < text.size()) return text[i];
  }
  throw Error("", "the keys table has no text column \"" + name + "\"");
}

void KeysTable::add_column(const std::string& name,
                           std::vector<double> values) {
  names.push_back(name);
  numeric.resize(names.size());
  text.resize(names.size());
  numeric.back() = std::move(values);
}

void KeysTable::add_text_column(const std::string& name,
                                std::vector<std::string> values) {
  names.push_back(name);
  numeric.resize(names.size());
  text.resize(names.size());
  text.back() = std::move(values);
}

void Prediction::check(const std::string& where) const {
  const std::string at = where.empty() ? std::string() : where + ": ";
  if (!uncertainty.has_value()) {
    if (level.has_value() || method.has_value()) {
      throw Error("", at + "level and method go with an uncertainty; a "
                           "prediction without one has neither (section 10)");
    }
    return;
  }
  if (uncertainty->shape != mean.shape) {
    throw Error("", at + "a band has the shape of its mean (section 10)");
  }
  if (!level.has_value() || !(*level > 0.0 && *level < 1.0)) {
    throw Error("", at + "a band states the coverage it claims as level in "
                         "(0, 1); a 1.96-sigma Gaussian band is 0.95");
  }
  if (!method.has_value() || method->empty()) {
    throw Error("", at + "a band says how it was made; give method one "
                         "sentence");
  }
  for (const double v : uncertainty->f64) {
    if (v < 0.0) {
      throw Error("", at + "a band is a half-width and is never negative");
    }
  }
}

bool Outputs::has(const std::string& output) const {
  return by_output.find(output) != by_output.end();
}

const Prediction& Outputs::at(const std::string& output) const {
  const auto it = by_output.find(output);
  if (it == by_output.end()) {
    throw Error("", "this callable has no output \"" + output + "\"");
  }
  return it->second;
}

void Outputs::set(const std::string& output, Prediction p) {
  by_output[output] = std::move(p);
}

namespace {

std::map<std::string, CallableRegistry::Factory, BytesLess>& registry() {
  static std::map<std::string, CallableRegistry::Factory, BytesLess> r;
  return r;
}

}  // namespace

void CallableRegistry::register_type(const std::string& type,
                                     Factory factory) {
  registry()[type] = std::move(factory);
}

bool CallableRegistry::knows(const std::string& type) {
  return registry().find(type) != registry().end();
}

std::unique_ptr<Callable> CallableRegistry::from_dict(
    const std::string& type, const Dict& dict) {
  const auto it = registry().find(type);
  if (it == registry().end()) return nullptr;
  return it->second(dict);
}

std::vector<std::string> CallableRegistry::types() {
  std::vector<std::string> out;
  for (const auto& entry : registry()) out.push_back(entry.first);
  return out;
}

}  // namespace mestra
